### PR TITLE
Add solution to Assignment 3

### PR DESCRIPTION
**Pull request by elifguler96**

### CODE REVIEW

File: Assignment3.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+        List<String> words = new ArrayList<>();
+        words.addAll(setOfWords);
+
+        return words;
+    }
+
+    private void findWordsStartingWithSpecificChar(char[][] grid, int row, int column, boolean[][] visited, String s,
+                                                   MyDictionary dictionary) {
+        if (!dictionary.isPrefix(s + grid[row][column])) {
```
Comment (katermar) [inline Assignment3.java]: Where is your class "MyDictionary"? Or it is supposed that it is given? How does the test acts without list of prefixes and list of words?

File: Assignment3.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            return;
+        }
+
+        visited[row][column] = true;
+        s += grid[row][column];
+
+        if (dictionary.isWord(s)) {
```
Comment (katermar) [inline Assignment3.java:53]: The same issue as higher mentioned
Or maybe you can add some comments 
(So, you need to add comments)

**Comment by katermar** [comment]
Please comment your code

**Comment by elifguler96** [comment]
You are right, I forgot 😅 I'll add the comments tonight 👍🏻 Thanks for the catch

**Comment by elifguler96** [comment]
I added the comments 😃

File: Assignment3.java [hunk 3]
```diff
@@ -20,8 +31,19 @@
         return words;
```
Comment (katermar) [inline Assignment3.java]: you can reduce this part of code just writing `return new ArrayList<String>(setOfWords);`

File: Assignment3.java [hunk 4]
```diff
@@ -3,13 +3,24 @@
 public class Assignment3 {
```
Comment (katermar) [inline Assignment3.java]: maybe I'm not right? but isn't it better to name your class smth like `WordsGridFinder` ? To make it more understandable .

**Review by katermar — COMMENTED** [review]
Thanks for adding comments, you've made it easier to understand. I've left a few comments, but almost everything looks nice for me.

File: Assignment3.java (hunk 4 above)
Comment (elifguler96) [reply Assignment3.java]: Yeah it is, but I just name all the assignments with their numbers, I made it sort of a convention :D

File: Assignment3.java (hunk 3 above)
Comment (elifguler96) [reply Assignment3.java]: Thanks 👍

File: Assignment3.java (hunk 1 above)
Comment (elifguler96) [reply Assignment3.java]: I have the Dictionary class, but as one of the mentors said in the slack channel and during the Hangout, I just used it for testing, I didn't submit the file.

screen shot 2017-06-14 at 10 21 25 am

Comment (elifguler96) [reply Assignment3.java]: Don't think of this as a Java file that is going to compile. Just think of it as you wrote this code in Google Docs, like in the interview. In the pdf, it says that the dictionary and its methods are given.

File: Assignment3.java (hunk 4 above)
Comment (katermar) [reply Assignment3.java]: Yeah, I've noticed that :)

File: Assignment3.java (hunk 1 above)
Comment (katermar) [reply Assignment3.java]: Oh, okay, I'm just curious how does your test without any prefix list and word list act :)

Comment (elifguler96) [reply Assignment3.java]: Of course, if you try to run the tests right now, it won't even compile. I run the tests with `MyDictionary` class and then submitted my code here.

File: Assignment3.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import java.util.*;
```
Comment (brzozan) [inline Assignment3.java]: Importing contents of whole package via * is generally discouraged. It pollutes local namespace and makes it non-obvious which classes are expected to be used.

See for example longer explanations on [Stack Overflow](https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad).

File: Assignment3.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+public class Assignment3 {
+    private Set<String> setOfWords = new HashSet<>();
```
Comment (brzozan) [inline Assignment3.java]: It would be a bit better from design perspective to pass this set as a parameter to the methods, so that instances of class Assignment3 are stateless. Currently if you call findAllWords twice with different grids, you will get incorrect result from the second call - list will contain words from both first and second grid, since first invocation changes the state of the object and modifies its behavior for future calls.

Managing state can be a source of serious and difficult to find bugs, that happen only after a specific sequence of events happens. A simple example for this case is shown above.
Storing state in the object should have good reason, for example caching of generated results so that the next invocation with the same parameters returns cached result rather than generate new one.

File: Assignment3Test.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.HashSet;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (brzozan) [inline Assignment3Test.java]: Importing lots of stuff just to use a single assertEquals :(

File: Assignment3Test.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+class Assignment3Test {
+    char[][] grid = {{'a', 'a', 'r'}, {'t', 'c', 'd'}};
+
+    @Test
+    void findAllWordsTest() {
+        assertEquals(new HashSet<>(new Assignment3().findAllWords(2, 3, grid, new MyDictionary())),
```
Comment (brzozan) [inline Assignment3Test.java]: Usual practice is to create tested object in a separate method annotated with `@Before` and only use it in test.

Also: clever use of wrapping in HashSets to work around changing order of elements! There are nice [libraries](https://google.github.io/truth/) that let you handle it in a cleaner way[1], but feel free to leave it as is to not overcomplicate things.

[1] `assertThat(result).containsExactly(s1, s2, s3);`

File: Assignment3Test.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.HashSet;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class Assignment3Test {
```
Comment (brzozan) [inline Assignment3Test.java]: Any tests for edge cases, like empty grid or grid with no words from dictionary?

File: Assignment3.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    *
+    * @param  grid       gird of characters
+    * @param  row        number of rows in the grid
+    * @param  column     number of columns in the grid
+    * @param  visited    holds whether a character is visited
+    * @param  s          forms the word
+    * @param  dictionary dictionary containing the words
+    */
+    private void formWords(char[][] grid, int row, int column, boolean[][] visited, String s,
+                                                   MyDictionary dictionary) {
```
Comment (brzozan) [inline Assignment3.java]: A minor thing, but makes code difficult to read: line with MyDictionary is indented waaaaaaaaaaaay too far.

**Review by brzozan — CHANGES_REQUESTED** [review]
Very good! Only a few comments.

File: Assignment3.java (hunk 5 above)
Comment (elifguler96) [reply Assignment3.java]: IntelliJ did that automatically I guess 😅

File: Assignment3Test.java (hunk 8 above)
Comment (elifguler96) [reply Assignment3Test.java]: I searched for that but couldn't find it, thanks 😄

**Review by brzozan — APPROVED** [review]
LGTM
